### PR TITLE
Pr82 followup

### PR DESCRIPTION
**Pull request by qianfengz**
This P.R is following P.R 82, it has the following added/implemented
1) Using cluster descriptor to get thread cluster IDs of `m` or `k` dimension, and use `2d M_K` tensor to describe the `Blockwise` LDS reduction buffer, making the codes simpler (`kernels`)
2) Change to all the reduction device interfaces to use `NumReduceDims` as template parameter to replace the old method of using `ReduceDims`,  this change will simply the numeration of the `add_device_reduce_xxx` method instances (`devices`)
3) Add Readme files for the pool2d_fwd and reduce_blockwise examples (`examples`)
4) Update to the Reduction testing scripts to make them aligned with the style of other test scripts (`test`)
5) Other possible updates in names

### CODE REVIEW

File: example/12_reduce/README.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
+# Instructions for ```conv2d_bwd_data_xdl``` Example
```
Comment (asroy) [inline example/12_reduce/README.md:1]: reduce_blockwise

File: example/12_reduce/reduce_blockwise.cpp [hunk 2]
```diff
@@ -28,8 +29,8 @@ using kInDataType  = ck::half_t;
 using kOutDataType = ck::half_t;
 using kAccDataType = float;
 
-constexpr int Rank = 4;
-using ReduceDims_  = ck::Sequence<0, 1, 2>;
+constexpr int Rank          = 4;
+constexpr int NumReduceDims = 3;
```
Comment (asroy) [inline example/12_reduce/reduce_blockwise.cpp:33]: NumReduceDim

```Xs``` stands for a container of ```X```

File: include/ck/tensor_operation/gpu/device/device_reduce_blockwise.hpp [hunk 3]
```diff
@@ -15,8 +15,8 @@ namespace device {
 template <typename InDataType,
           typename AccDataType,
           typename OutDataType,
-          int Rank,
-          typename ReduceDims,
+          index_t Rank,
+          index_t NumReduceDims,
```
Comment (asroy) [inline include/ck/tensor_operation/gpu/device/device_reduce_blockwise.hpp:19]: NumReduceDim

File: include/ck/tensor_operation/gpu/device/device_reduce_blockwise.hpp [hunk 4]
```diff
@@ -148,26 +154,27 @@ struct DeviceReduceBlockWise : public DeviceReduce<InElementwiseOperation, AccEl
...
 
-            inLengths_  = inLengths;
-            inStrides_  = inStrides;
             outLengths_ = outLengths;
             outStrides_ = outStrides;
 
+            std::tie(inLengths_, inStrides_) =
+                shuffle_tensor_dimensions<Rank, NumReduceDims>(inLengths, inStrides, toReduceDims);
+
             in_elementwise_op_  = in_elementwise_op;
```
Comment (asroy) [inline include/ck/tensor_operation/gpu/device/device_reduce_blockwise.hpp:163]: Move this into member initializer list

File: include/ck/tensor_operation/gpu/device/device_reduce_blockwise.hpp [hunk 5]
```diff
@@ -136,6 +141,7 @@ struct DeviceReduceBlockWise : public DeviceReduce<InElementwiseOperation, AccEl
                  const std::vector<int>& inStrides,
                  const std::vector<int>& outLengths,
                  const std::vector<int>& outStrides,
+                 const std::vector<int>& toReduceDims,
```
Comment (asroy) [inline include/ck/tensor_operation/gpu/device/device_reduce_blockwise.hpp:144]: inReduceDims

File: include/ck/tensor_operation/gpu/device/device_reduce_multiblock_atomic_add.hpp [hunk 6]
```diff
@@ -17,8 +17,8 @@ namespace device {
 template <typename InDataType,
           typename AccDataType,
           typename OutDataType,
-          int Rank,
-          typename ReduceDims,
+          index_t Rank,
+          index_t NumReduceDims,
```
Comment (asroy) [inline include/ck/tensor_operation/gpu/device/device_reduce_multiblock_atomic_add.hpp:21]: NumReduceDim

File: include/ck/tensor_operation/gpu/device/device_reduce_multiblock_atomic_add.hpp [hunk 7]
```diff
@@ -41,7 +41,12 @@ struct DeviceReduceMultiBlockAtomicAdd
 
     using IndexDataType = int32_t;
 
-    using InvariantDims = decltype(get_invariant_dims<Rank, ReduceDims>());
+    static constexpr index_t NumInvariantDims = Rank - NumReduceDims;
```
Comment (asroy) [inline include/ck/tensor_operation/gpu/device/device_reduce_multiblock_atomic_add.hpp:44]: NumInvariantDim

File: include/ck/tensor_operation/gpu/device/device_reduce_multiblock_atomic_add.hpp [hunk 8]
```diff
@@ -147,6 +152,7 @@ struct DeviceReduceMultiBlockAtomicAdd
                  const std::vector<int>& inStrides,
                  const std::vector<int>& outLengths,
                  const std::vector<int>& outStrides,
+                 const std::vector<int>& toReduceDims,
```
Comment (asroy) [inline include/ck/tensor_operation/gpu/device/device_reduce_multiblock_atomic_add.hpp:155]: inReduceDims

File: include/ck/tensor_operation/gpu/device/device_reduce_multiblock_atomic_add.hpp [hunk 9]
```diff
@@ -369,6 +376,7 @@ struct DeviceReduceMultiBlockAtomicAdd
                         const std::vector<int>& inStrides,
                         const std::vector<int>& outLengths,
                         const std::vector<int>& outStrides,
+                        const std::vector<int>& toReduceDims,
```
Comment (asroy) [inline include/ck/tensor_operation/gpu/device/device_reduce_multiblock_atomic_add.hpp:379]: inRediceDims

File: library/include/ck/library/tensor_operation_instance/gpu/reduce/device_reduce_instance_threadwise.hpp [hunk 10]
```diff
@@ -57,7 +57,7 @@ template <typename InDataType,
           typename AccDataType,
           typename OutDataType,
           int Rank,
-          typename ReduceDims,
+          int NumReduceDims,
```
Comment (asroy) [inline library/include/ck/library/tensor_operation_instance/gpu/reduce/device_reduce_instance_threadwise.hpp:60]: NumReduceDim

File: profiler/include/profile_reduce_impl.hpp [hunk 11]
```diff
@@ -9,54 +9,52 @@ namespace tensor_operation {
 namespace device {
 namespace device_reduce_instance {
 
-template <int Rank, typename ReduceDims, int ReduceOpId, int NanOpt, int IndicesOpt>
+template <int Rank, int NumReduceDims, int ReduceOpId, int NanOpt, int IndicesOpt>
```
Comment (asroy) [inline profiler/include/profile_reduce_impl.hpp:12]: NumReduceDim

File: profiler/include/profile_reduce_impl.hpp [hunk 12]
```diff
@@ -9,54 +9,52 @@ namespace tensor_operation {
...
 struct ReduceDescription
 {
-    static constexpr int Rank_       = Rank;
-    static constexpr int ReduceOpId_ = ReduceOpId;
-    static constexpr int NanOpt_     = NanOpt;
-    static constexpr int IndicesOpt_ = IndicesOpt;
-
-    using ReduceDims_ = ReduceDims;
+    static constexpr int Rank_          = Rank;
+    static constexpr int NumReduceDims_ = NumReduceDims;
```
Comment (asroy) [inline profiler/include/profile_reduce_impl.hpp:16]: NumReduceDim_

File: profiler/include/profile_reduce_impl.hpp [hunk 13]
```diff
@@ -66,16 +64,11 @@ bool description_match(const DescriptionType& description,
        description.IndicesOpt_ != static_cast<int>(IndicesOpt))
         return (false);
 
-    if(DescriptionType::ReduceDims_::Size() != ReduceDims.size())
+    if(DescriptionType::NumReduceDims_ != toReduceDims.size())
```
Comment (asroy) [inline profiler/include/profile_reduce_impl.hpp:67]: name

File: profiler/include/profile_reduce_impl.hpp [hunk 14]
```diff
@@ -584,7 +576,7 @@ void profile_reduce_impl(bool do_verification,
                          bool do_dumpout,
                          int nrepeat,
                          const std::vector<size_t>& inLengths,
-                         const std::vector<int>& ReduceDims,
+                         const std::vector<int>& toReduceDims,
```
Comment (asroy) [inline profiler/include/profile_reduce_impl.hpp:579]: inReduceDims or reduceDims

File: profiler/include/profile_reduce_impl.hpp [hunk 15]
```diff
@@ -605,18 +597,26 @@ void profile_reduce_impl(bool do_verification,
         using descType = remove_cvref_t<decltype(std::get<i>(tuple_object))>;
 
         if(!description_match(
-               descType{}, inLengths.size(), ReduceDims, ReduceOpId, NanOpt, IndicesOpt))
+               descType{}, inLengths.size(), toReduceDims, ReduceOpId, NanOpt, IndicesOpt))
```
Comment (asroy) [inline profiler/include/profile_reduce_impl.hpp:600]: name

File: profiler/include/profile_reduce_impl.hpp [hunk 16]
```diff
@@ -605,18 +597,26 @@ void profile_reduce_impl(bool do_verification,
...
             return;
 
         profile_reduce_impl_impl<InDataType,
                                  AccDataType,
                                  OutDataType,
                                  descType::Rank_,
-                                 typename descType::ReduceDims_,
+                                 descType::NumReduceDims_,
```
Comment (asroy) [inline profiler/include/profile_reduce_impl.hpp:607]: name

**Review by asroy — CHANGES_REQUESTED** [review]
Please fix naming

File: include/ck/tensor_operation/gpu/device/device_reduce_blockwise.hpp (hunk 4 above)
Comment (qianfengz) [reply include/ck/tensor_operation/gpu/device/device_reduce_blockwise.hpp:163]: Do you mean move `outLengths_ = outLengths; outStrides_ = outStrides;` and `in_elementwise_op_  = in_elementwise_op; acc_elementwise_op_ = acc_elementwise_op` to the member initializer list ?

**Comment by qianfengz** [comment]
All required fixing has been done!

File: profiler/include/profile_reduce_impl.hpp (hunk 16 above)
Comment (qianfengz) [reply profiler/include/profile_reduce_impl.hpp]: done

File: profiler/include/profile_reduce_impl.hpp (hunk 15 above)
Comment (qianfengz) [reply profiler/include/profile_reduce_impl.hpp]: done

File: profiler/include/profile_reduce_impl.hpp (hunk 14 above)
Comment (qianfengz) [reply profiler/include/profile_reduce_impl.hpp]: done

File: profiler/include/profile_reduce_impl.hpp (hunk 13 above)
Comment (qianfengz) [reply profiler/include/profile_reduce_impl.hpp]: done

File: profiler/include/profile_reduce_impl.hpp (hunk 12 above)
Comment (qianfengz) [reply profiler/include/profile_reduce_impl.hpp]: done

File: library/include/ck/library/tensor_operation_instance/gpu/reduce/device_reduce_instance_multiblock_atomic_add.hpp [hunk 17]
```diff
@@ -132,37 +132,37 @@ void add_device_reduce_instance_multiblock_atomic_add(
...
-                                                                   compT,                 \
-                                                                   outT,                  \
-                                                                   Rank,                  \
-                                                                   Sequence<__VA_ARGS__>, \
-                                                                   ReduceOpId,            \
-                                                                   NanOpt,                \
-                                                                   IndicesOpt>(           \
-        std::vector<deviceReduceMultiBlockAtomicAddPtrType<compT, ReduceOpId>> &          \
+#define ADD_MULTIBLOCK_ATOMIC_ADD_INST_BY_TYPE(                                   \
+    inT, compT, outT, ReduceOpId, NanOpt, IndicesOpt, Rank, NumReduceDims)        \
```
[the remark this answers is not in the document]
Comment (qianfengz) [reply library/include/ck/library/tensor_operation_instance/gpu/reduce/device_reduce_instance_multiblock_atomic_add.hpp:136]: done

File: library/include/ck/library/tensor_operation_instance/gpu/reduce/device_reduce_instance_blockwise.hpp [hunk 18]
```diff
@@ -57,7 +57,7 @@ template <typename InDataType,
           typename AccDataType,
           typename OutDataType,
           int Rank,
-          typename ReduceDims,
+          int NumReduceDims,
```
Comment (asroy) [inline library/include/ck/library/tensor_operation_instance/gpu/reduce/device_reduce_instance_blockwise.hpp:60]: NumReduceDim

File: library/include/ck/library/tensor_operation_instance/gpu/reduce/device_reduce_instance_blockwise.hpp [hunk 19]
```diff
@@ -149,15 +151,16 @@ void add_device_reduce_instance_blockwise(
...
+#define ADD_BLOCKWISE_INST_REF_BY_ID(                                              \
+    inT, compT, outT, ReduceOpId, NanOpt, IndicesOpt, Rank, NumReduceDims)         \
+    ADD_BLOCKWISE_INST_REF_BY_TYPE(inT,                                            \
+                                   compT,                                          \
+                                   outT,                                           \
+                                   static_cast<ReduceTensorOp_t>(ReduceOpId),      \
+                                   static_cast<NanPropagation_t>(NanOpt),          \
+                                   static_cast<ReduceTensorIndices_t>(IndicesOpt), \
+                                   Rank,                                           \
+                                   NumReduceDims)
```
Comment (asroy) [inline library/include/ck/library/tensor_operation_instance/gpu/reduce/device_reduce_instance_blockwise.hpp:163]: NumReduceDims

File: library/include/ck/library/tensor_operation_instance/gpu/reduce/device_reduce_instance_blockwise_second_call.hpp [hunk 20]
```diff
@@ -45,7 +45,7 @@ template <typename InDataType,
           typename AccDataType,
           typename OutDataType,
           int Rank,
-          typename ReduceDims,
+          int NumReduceDims,
```
Comment (asroy) [inline library/include/ck/library/tensor_operation_instance/gpu/reduce/device_reduce_instance_blockwise_second_call.hpp:48]: NumReduceDims

File: library/include/ck/library/tensor_operation_instance/gpu/reduce/device_reduce_instance_multiblock_atomic_add.hpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (asroy) [reply library/include/ck/library/tensor_operation_instance/gpu/reduce/device_reduce_instance_multiblock_atomic_add.hpp:136]: still not fixed

File: library/include/ck/library/tensor_operation_instance/gpu/reduce/device_reduce_instance_multiblock_atomic_add.hpp [hunk 21]
```diff
@@ -132,37 +132,37 @@ void add_device_reduce_instance_multiblock_atomic_add(
...
+    template void add_device_reduce_instance_multiblock_atomic_add<inT,           \
+                                                                   compT,         \
+                                                                   outT,          \
+                                                                   Rank,          \
+                                                                   NumReduceDims, \
```
Comment (asroy) [inline library/include/ck/library/tensor_operation_instance/gpu/reduce/device_reduce_instance_multiblock_atomic_add.hpp:141]: NumReduceDim

File: library/include/ck/library/tensor_operation_instance/gpu/reduce/device_reduce_instance_multiblock_atomic_add.hpp [hunk 22]
```diff
@@ -132,37 +132,37 @@ void add_device_reduce_instance_multiblock_atomic_add(
...
+    inT, compT, outT, ReduceOpId, NanOpt, IndicesOpt, Rank, NumReduceDims)                 \
     ADD_MULTIBLOCK_ATOMIC_ADD_INST_BY_TYPE(inT,                                            \
                                            compT,                                          \
                                            outT,                                           \
                                            static_cast<ReduceTensorOp_t>(ReduceOpId),      \
                                            static_cast<NanPropagation_t>(NanOpt),          \
                                            static_cast<ReduceTensorIndices_t>(IndicesOpt), \
                                            Rank,                                           \
-                                           __VA_ARGS__)
+                                           NumReduceDims)
```
Comment (asroy) [inline library/include/ck/library/tensor_operation_instance/gpu/reduce/device_reduce_instance_multiblock_atomic_add.hpp:157]: name

**Comment by asroy** [comment]
Please do a grep and replace in all files

NumReduceDims -> NumReduceDim

**Review by asroy — APPROVED** [review]

**Comment by qianfengz** [comment]
> Please do a grep and replace in all files
> 
> NumReduceDims -> NumReduceDim

Done !  Thanks

**Review by asroy — APPROVED** [review]

**Pull request merged**